src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -1195,6 +1268,186 @@ static status_t del_routes_internal(private_kernel_super_ipsec_t* this,
+						  chunk_hash(chunk_from_str(tunnel_key->vrf_name)))));
+}
+
+
+static void destroy_tunnel_counter_key(void *value, const void *key) {
+	(void)key;
+	if(! value) {
+		return;
+	}
+
+	struct tunnel_counter_key *tunnel_key = (struct tunnel_counter_key *)value;
+	free(tunnel_key->tenant_id);
+	free(tunnel_key->account_id);
+	free(tunnel_key->vrf_name);
+	free(value);
+}
+
+CALLBACK(write_terminator_tenant_account_vrf_tunnels, void,
+		 void *ctx, const struct system_enterprise_map_iterator* iter) {
+	hashtable_t *counter_map = (hashtable_t *)ctx;
+	const char* const tenant_id = iter->lookup_params->tenant_id;
+	const char* const account_id = iter->lookup_params->account_id;
+	const char* const vrf_name = iter->lookup_params->vrf_name;
+	uint32_t term_id = iter->lookup_params->terminator_id;
+
+	struct tunnel_counter_key key = {.terminator_id = term_id,
+									 .tenant_id = tenant_id,
+									 .account_id = account_id,
+									 .vrf_name = vrf_name,
+									 .count = 0};
+	struct tunnel_counter_key *curr_key = counter_map->get(counter_map, &key);
+	if(NULL != curr_key) {
+		key.count = curr_key->count;
+	}
+
+	if (! oper_db_writer_set_terminator_vrf_tunnel_counter(term_id,
+														   tenant_id, account_id, vrf_name,
+														   key.count)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "failed writing vrf: %s in term %u tunnel counter",
+			vrf_name, curr_key->terminator_id);
+	}
+}
+
+CALLBACK(write_terminator_tunnels_oper, void,
+		private_kernel_super_ipsec_t* this,
+		struct pm_terminator_conn* const terminator_conn) {
+	const struct pm_proto_addr* const lookup_addr = picomsg_terminator_conn_get_lookup_addr(terminator_conn);
+	enumerator_t* enumerator = NULL;
+	ike_sa_t* ike_sa;
+	uint64_t term_tunnels_number = 0;
+	hashtable_t *counter_map = NULL;
+	const uint32_t term_id = picomsg_terminator_conn_get_id(terminator_conn);
+
+	counter_map = hashtable_create(tunnel_counter_key_hash, tunnel_counter_key_compare, 10);
+
+	enumerator = charon->controller->create_ike_sa_enumerator(charon->controller, true);
+	while (enumerator->enumerate(enumerator, &ike_sa)) {
+		struct pm_proto_addr ike_sa_addr;
+		host_addr_to_picomsg_addr(ike_sa->get_my_host(ike_sa), &ike_sa_addr);
+		if (! pm_proto_addr_equals(&ike_sa_addr, lookup_addr)) {
+			continue;
+		}
+
+		const ike_super_peer_info_t* const super_peer_info = ike_sa->get_super_peer_info(ike_sa);
+		if (! super_peer_info) {
+			continue;
+		}
+
+		if (IKE_SUPER_PEER_INFO_TYPE_TERM != super_peer_info->get_type(super_peer_info)) {
+			continue;
+		}
+
+		struct tunnel_counter_key *key;
+		INIT(key,
+			 .terminator_id = term_id,
+			 .tenant_id = strdupnull(super_peer_info->get_tenant_id(super_peer_info)),
+			 .account_id = strdupnull(super_peer_info->get_account_id(super_peer_info)),
+			 .vrf_name = strdupnull(super_peer_info->get_vrf_name(super_peer_info)),
+			 /** 'count' is not used for comparison */
+		);
+		if(! key || ! key->tenant_id || ! key->account_id || ! key->vrf_name) {
```

Comment:
The tenant_id, account_id and vrf_name will never be NULL in super_peer_info so these checks are redundant (if you're worried about malloc() failing then not only handling this gracefully is useless (if you fail to malloc say 50 bytes then the program is in such a bad state it will likely crash soon anyways). And on top of that the whole code base doesn't check for malloc failing (for the above reason presumably), so adding such checks here will not only not buy us anything (even if you handle malloc(50) failing then the rest of the code doesn't handle, and so we'll crash there), it also complicates the code by adding useless checks.